File: MetodyObliczeniowe/MetodyObliczeniowe/Wielomian.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MetodyObliczeniowe
{
	public class Wielomian : ICloneable
	{
		#region Public properites

		//P(x)
		public string PodX { get; set; }
		// a0 + a1x0 +a2x0^2 + a3x0^3
		public List<Wezel> RownanieWielomianu;
		public List<Wezel> CalkowanyWielominan;

		#endregion

		#region public method

		public Wielomian(string peOdX, int iloscWspolczynnikow)
		{
			PodX = peOdX;
			RownanieWielomianu = new List<Wezel>(iloscWspolczynnikow);
			for (int i = 0; i < iloscWspolczynnikow; i++)
			{
				RownanieWielomianu.Add(new Wezel(i, peOdX));
			}
		}

		public void PostacOgolna()
		{
			PodX = "x";
			foreach (var wezel in RownanieWielomianu)
			{
				wezel.X = "x";
			}
		}

		public override string ToString()
		{
			StringBuilder sb = new StringBuilder();
			sb.Append("P(" + PodX + ") = ");
			var rownaniePomocnicze = new List<Wezel>(RownanieWielomianu);
			sb.Append(rownaniePomocnicze.First().ToString());
			rownaniePomocnicze.Remove(rownaniePomocnicze.First());
			rownaniePomocnicze.ForEach(wezel =>
			{
				string textWezel = wezel.ToString();
				if (!string.IsNullOrEmpty(textWezel))
					sb.Append(string.Format(" + {0}", textWezel));
			});
			return sb.ToString();
		}

		public void PodstawAn(Dictionary<string, double> Ans)
		{
			RownanieWielomianu.ForEach(wezel =>
			{
				if (Ans.Keys.Contains(wezel.Wspolczynnik))
				{
					wezel.Wspolczynnik = Ans[wezel.Wspolczynnik].ToString();
				}
			});
		}

		public void WyliczGdzieSieDa()
		{
			RownanieWielomianu.ForEach(wezel =>
				{
					if (wezel.ToString().Contains("a")) return;
					wezel.WartoscWezla = Convert.ToDouble(wezel.WspolczynnikPomocniczy) * Convert.ToDouble(wezel.Wspolczynnik) * Math.Pow(Convert.ToDouble(wezel.X), Convert.ToDouble(wezel.Potega));
				});
		}

		public double PoliczCalke(double a, double b)
		{
			return b > a ? Math.Round(PoliczWartoscWielomianuWPunkcie(b) - PoliczWartoscWielomianuWPunkcie(a), 4) : Math.Round(PoliczWartoscWielomianuWPunkcie(a) - PoliczWartoscWielomianuWPunkcie(b), 4);
		}

		public void WyznaczCalke()
		{
			CalkowanyWielominan = new List<Wezel>();

			foreach (var wezel in RownanieWielomianu)
			{
				double wspolczynnik;
				double.TryParse(wezel.Wspolczynnik, out wspolczynnik);
				int newPotega = wezel.Potega + 1;
				wspolczynnik /= newPotega;
				CalkowanyWielominan.Add(new Wezel(newPotega, wspolczynnik.ToString(), wezel.WspolczynnikPomocniczy));
			}
		}

		public string An()
		{
			return RownanieWielomianu.First().Wspolczynnik;
		}

		public double ToCoStoiPrzyA()
		{
			return RownanieWielomianu.First().WspolczynnikPomocniczy;
		}

		public object Clone()
		{
			return this.MemberwiseClone();
		}

		#endregion

		#region private method

		private double PoliczWartoscWielomianuWPunkcie(double p)
		{
			double suma = 0.0;
			foreach (var wezel in CalkowanyWielominan)
			{
				double wspolczynnik;
				Double.TryParse(wezel.Wspolczynnik, out wspolczynnik);
				suma += wspolczynnik * wezel.WspolczynnikPomocniczy * Math.Pow(p, wezel.Potega);
			}
			return suma;
		}

		#endregion
	}
}

File: zadanie_1/lib/Wielomian.rb
require_relative './Wezel'

class Wielomian
  attr_accessor :PodX
  attr_accessor :RownanieWielomianu
  attr_accessor :CalkowanyWielominan

  def initialize(peOdX, iloscWspolczynnikow)
    if(IsValidParameters(peOdX, iloscWspolczynnikow))
      @PodX = peOdX
      @RownanieWielomianu = Array.new
      for i in 0..iloscWspolczynnikow - 1
        @RownanieWielomianu << Wezel.new(i, peOdX)
      end
    else
      raise "Invalid argument"
    end
  end

  def PostacOgolna()  
    @PodX = "x"
    @RownanieWielomianu.each { |wezel| 
      wezel.X = "x"
    } 
  end

  def ToString()
      sb = String.new
      sb += "P(#{@PodX}) = "

      rownaniePomocnicze = Array.new
      @RownanieWielomianu.each { |wezel| rownaniePomocnicze << wezel}
      sb += rownaniePomocnicze.first.ToString
      rownaniePomocnicze.shift

      rownaniePomocnicze.each { |wezel| 
        txt = wezel.ToString
        if (!txt.empty?)
          sb += " + #{txt}"
        end
      }

      return sb
  end

  # podstawia wartosci zamiast an 
  def PodstawAn(ans)  
    @RownanieWielomianu.each{ |wezel|  
      if(ans.key?(wezel.Wspolczynnik))
        wezel.Wspolczynnik = ans[wezel.Wspolczynnik]
      end
    }
  end
  
  def WyliczGdzieSieDa()
    @RownanieWielomianu.each { |wezel|
      if(!wezel.ToString.include? "a")
        wezel.WartoscWezla = wezel.Wspolczynnik.to_f * wezel.WspolczynnikPomocniczy.to_f * (wezel.X.to_f ** wezel.Potega.to_f)
      end
    }
  end

  def PoliczCalke(a, b)
    if(b > a)
      return (PoliczWartoscWielomianuWPunkcie(b) - PoliczWartoscWielomianuWPunkcie(a)).round(4)
    else
      return (PoliczWartoscWielomianuWPunkcie(a) - PoliczWartoscWielomianuWPunkcie(b)).round(4)
    end
  end
  
  def WyznaczCalke()
    @CalkowanyWielominan = Array.new
    @RownanieWielomianu.each { |wezel| 
      wspolczynnik = wezel.Wspolczynnik.to_f
      nowaPotega = wezel.Potega + 1
      wspolczynnik = wspolczynnik / nowaPotega
      @CalkowanyWielominan << Wezel.new(nowaPotega, "x", wspolczynnik.to_s, wezel.WspolczynnikPomocniczy)
    }
  end

  # zwraca wartosc wspolczynnika ktory nie zostal jeszcze obliczony, tzn zwraca an..a1
  def An()
    return @RownanieWielomianu.first.Wspolczynnik
  end
  
  # zwraca wartosc tego co stoi przy an..a1, jako wspolczynnik pomocniczy
  def ToCoStoiPrzyA()
    return @RownanieWielomianu.first.WspolczynnikPomocniczy
  end

  private
 
  def IsValidParameters(peOdX, iloscWspolczynnikow)
    isValid = true
  
    if(!peOdX.is_a? String)
      isValid = false
    end

    if(!iloscWspolczynnikow.is_a? Fixnum)
      isValid = false
    end

    return isValid
  end

  def PoliczWartoscWielomianuWPunkcie(p)
    suma = 0.0;
      @CalkowanyWielominan.each { |wezel|
        wspolczynnik = wezel.Wspolczynnik.to_f
        suma += wspolczynnik * wezel.WspolczynnikPomocniczy * (p ** wezel.Potega)
      }
    return suma
  end
end